<a href="https://colab.research.google.com/github/iolopezm/Analitica-de-Negocios/blob/main/RETO1_IOLMipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np #Funciones Matemáticas
import pandas as pd #Comunicación con Excel
from sklearn.naive_bayes import GaussianNB #Función Gaussiana
from sklearn.metrics import confusion_matrix #La métrica de Medida

In [ ]:
XDB.columns

Index(['ID', 'Age', 'Urban', 'Mother_Edu', 'Father_Edu', 'Siblings', 'Black',
       'Hispanic', 'White', 'Christian', 'WomenPlace', 'Male', 'FamilySize',
       'Self_Esteem', 'Height', 'Weight', 'Outgoing_Kid', 'Outgoing_Adult',
       'HealthPlan', 'Income', 'Marital_Status', 'Education', 'WeeksEmployed',
       'NumberSpouses'],
      dtype='object')

In [ ]:
nxl="/content/sample_data/1. BD1_Longitudinal_Survey_Int.xlsx"
XDB=pd.read_excel(nxl,sheet_name=0).dropna() #Traigo los datos


#Seleccionamos las variables de interes
XDB=XDB[['Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem','FamilySize', 'Urban']]
XDB.head()

#Datos de Entrada -np.array() convertir a numeros
XD=np.array(XDB[['Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem','FamilySize']])

#Datos de Salida - Valor Deseado - que quiero que aprenda el modelo
yd=np.array(XDB['Urban'])

**¿Cuál es el numero de datos total de la base de datos?**

r// 5683 datos

In [ ]:
total_datos = XDB.shape[0]
total_datos

5683

In [ ]:
total_datos_urbanos = XDB[XDB['Urban'] == 1].shape[0]
total_datos_urbanos

4510

**Qué proporción del total de los datos corresponde al area rural?**

In [ ]:
total_datos_rural = XDB[XDB['Urban'] == 0].shape[0]
total_datos_rural

1173

In [ ]:
tasa_rural = total_datos_rural /total_datos
tasa_rural

0.20640506774590886

**Que proporcion del total de los datos corresponde al area urbana?**

In [ ]:
tasa_urbanos = total_datos_urbanos / total_datos
tasa_urbanos

0.7935949322540912

In [ ]:
mnb=GaussianNB() #Llamamos el modelo
mnb.fit(XD,yd)
ydp=mnb.predict(XD) #Muestra los datos de salida - pero que el obtuvo

#Parametros del Modelo por categoría - PreApr/PreNeg
u=mnb.theta_ #Se estima las medias
sigma=np.sqrt(mnb.var_) #Se estima las varianzas

#Las variaciones de las variables
LI=u-sigma
LS=u+sigma

print("La media de las variables por categoría:")
print(u)
print("El limite inferior de la variacion de las variables por categoria:")
print(LI)
print("El limite superior de la variacion de las variables por categoría:")
print(LS)

La media de las variables por categoría:
[[1.83810742e+01 1.07612958e+01 1.03324808e+01 6.72148338e+01
  1.48857630e+02 3.03672506e+04 2.20272805e+01 4.74168798e+00]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 6.69031042e+01
  1.44760976e+02 3.34278929e+04 2.26004435e+01 4.68780488e+00]]
El limite inferior de la variacion de las variables por categoria:
[[ 1.58996876e+01  7.58892579e+00  6.46311334e+00  6.31369799e+01
   1.18746964e+02 -1.33465823e+03  1.78344685e+01  2.16621834e+00]
 [ 1.58811112e+01  7.79013864e+00  7.15097738e+00  6.26897933e+01
   1.14669654e+02 -2.69533531e+03  1.84154526e+01  2.24996381e+00]]
El limite superior de la variacion de las variables por categoría:
[[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.12926877e+01
  1.78968296e+02 6.20691595e+04 2.62200925e+01 7.31715762e+00]
 [2.08328577e+01 1.45637416e+01 1.53621047e+01 7.11164151e+01
  1.74852297e+02 6.95511211e+04 2.67854344e+01 7.12564594e+00]]
La matriz de confusión:
[[   2 1171]
 [   4 4506]]


**Determinar el intervalo en el cual se encuentra la variación de la Edad**

AREA URBANA: El límite inferior es 1.58996876e+01 y el límite superior 2.08624608e+01

AREA RURAL:  El límite inferior es 1.58811112e+01 y el límite superior 2.08328577e+01

**Determinar el intervalo en el cual se encuentra la variación de la Altura**

AREA URBANA:  El límite inferior es 6.31369799e+01 y el límite superior es 7.12926877e+01

AREA RURAL: El límite inferior es 6.26897933e+01 y el límite superior es 7.11164151e+01

**Determinar el intervalo en el cual se encuentra la variación del Peso de las Personas**

AREA URBANA: El límite inferior es 1.48857630e+02 y el límite superior es 1.78968296e+02

AREA RURAL: El límite inferior es 1.14669654e+0 y el límite superior es 1.74852297e+02

In [ ]:
VN=cm[0,0] #Verdadero Negativo
FP=cm[0,1] #Falso Positivo
FN=cm[1,0] #Falso Negativo
VP=cm[1,1] #Verdadero Positivo

Exactitud=(VP+VN)/len(XD)
print("La exactitud del modelo es:",Exactitud) #Comportamiento General

TE=(FN+FP)/len(XD)
print("La tasa de error del modelo es:",TE) #Porcentaje Equivocaciones

Sensibilidad=VP/(VP+FP)
print("La sensibilidad es:",Sensibilidad) #Como se comportó frente a los positivos

Especificidad=VN/(VN+FN)
print("La especificidad es:",Especificidad) #Como se comportó frente a los negativos


La exactitud del modelo es: 0.7932430054548654
La tasa de error del modelo es: 0.2067569945451346
La sensibilidad es: 0.793729082261758
La especificidad es: 0.3333333333333333


**¿Para una persona con las siguientes características indicar si vive en la ciudad o las áreas rurales 34,9,6,65,135,50000,30,3?**

In [ ]:
Xun=np.array([[34,9,6,65,135,50000,30,3]]) #Esta es la persona del Excel
ydun=mnb.predict(Xun)
print(ydun)

[1.]


Esta persona vive en una zona urbana.

**Analísis**

Al ver una tasa de exactitud de 0.79 y una tasa de error de 0.21, el modelo tiene un buen rendimiento general- pero no excelente. Al ver una alta sensibilidad y una especificidad baja, se puede decir que el modelo es bueno indentificando a las personas que viven en zonas urbanas.

En resumen, el modelo funciona bien en general, prediciendo correctamente la mayoría de los casos y siendo especialmente preciso al identificar a las personas que viven en zonas urbanas. Sin embargo, se puede mejorar identificando de las personas que viven en zonas rurales. Al mejorar este aspecto, se podría reducir la tasa de errores del modelo y aumentar su efectividad general.